cellxgene_schema_cli/cellxgene_schema/ontology_processing.py

```diff
@@ -0,0 +1,192 @@
+        if download_format == "gz":
+            urllib.request.urlretrieve(url, output_file + ".gz")
+            _decompress(output_file + ".gz", output_file)
+            os.remove(output_file + ".gz")
+        else:
+            urllib.request.urlretrieve(url, output_file)
+
+
+def _decompress(infile, tofile):
+    """
+    Decompresses a gziped file
+    """
+    with open(infile, "rb") as inf, open(tofile, "w", encoding="utf8") as tof:
+        decom_str = gzip.decompress(inf.read()).decode("utf-8")
+        tof.write(decom_str)
+
+
+def _parse_owls(
+    working_dir=os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "ontology_files"
+    ),
+    owl_info_yml=os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "ontology_files/owl_info.yml"
+    ),
+    output_json_file=os.path.join(
+        os.path.dirname(os.path.realpath(__file__)),
+        "ontology_files/all_ontology.json.gz",
+    ),
+):
+
+    """
+    Parser all owl files in working_dir. Extracts information from all classes in the owl file.
+    The extracted information is written into a gzipped a json file with the following structure:
+    {
+        "ontology_name":
+            {
+            "term_id": {
+                "label": "..."
+                "ancestors": [
+                    "ancestor1_term_id_1",
+                    "ancestor2_term_id_2"
+                    ]
+                }
+            }
+
+            "term_id2": {
+                ...
+            }
+
+            ...
+            }
+    }
+    """
+
+    with open(owl_info_yml, "r") as owl_info_handle:
+        owl_info = yaml.safe_load(owl_info_handle)
+
+    owl_files = [
+        os.path.join(working_dir, i)
+        for i in os.listdir(working_dir)
+        if i.endswith(".owl")
+    ]
+
+    # Parse owl files
+    onto_dict = {}
+    for i in owl_files:
+        onto = owlready2.get_ontology(i).load()
+        onto_dict[onto.name] = {}
+
+        print(f"Processing {onto.name}")
+        for onto_class in onto.classes():
+
+            term_id = onto_class.name.replace("_", ":")
+
+            # Skip terms that are not direct children from this ontology
+            if not onto.name == term_id.split(":")[0]:
+                continue
+
+            # If there are specified target terms then only work with them
+            if onto.name in owl_info:
+                if "only" in owl_info[onto.name]:
+                    if term_id not in owl_info[onto.name]["only"]:
+                        continue
+
+            onto_dict[onto.name][term_id] = dict()
+            try:
+                onto_dict[onto.name][term_id]["label"] = onto_class.label[0]
+            except IndexError:
+                onto_dict[onto.name][term_id]["label"] = ""
+
+            onto_dict[onto.name][term_id]["ancestors"] = [
+                i.name.replace("_", ":")
+                for i in onto_class.ancestors()
+                if i.name.split("_")[0] == onto.name
```

Comment:
We may need to look into this further to ensure that the ordering is returned correctly. For example:

{'label': 'lung neoplasm',
 'ancestors': ['MONDO:0024505',
  'MONDO:0000001',
  'MONDO:0005275',
  'MONDO:0005087',
  'MONDO:0045024',
  'MONDO:0000651',
  'MONDO:0021117',
  'MONDO:0021350',
  'MONDO:0020641',
  'MONDO:0000270',
  'MONDO:0005070',
  'MONDO:0023370',
  'MONDO:0021199']}

It’s odd to me that 0024505 appears before 0000001 since it’s a child of 0000001?

https://www.ebi.ac.uk/ols/ontologies/mondo/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FMONDO_0024505

Reply:
https://www.ebi.ac.uk/ols/docs/neo4j-schema

> OLS provides a way to represent an OWL ontology as a simple graph in Neo4j. OWL ontologies aren’t graphs, but rather a collection of axioms. However, graph based serialisations of OWL exists, such as the the Resource Description Framework (RDF) serilaisation of OWL. The RDF serialisation of OWL requires extensive reification and uses blank nodes in order to represent the OWL axioms as a graph. This can make working with OWL in RDF rather cumbersome, but it does at least provide a complete description of OWL in RDF. If you want to work with the OWL RDF representation we recommend loading OWL RDF files directly into a dedicated RDF databse (such as Virtuoso), where you will be able to query the RDF graph with SPARQL.

Reply:
This is interesting, they are definitely not ordered and were stored as fetched by owl_ready2. I don't think this is something that is worth exploring for this specific feature as the ordering doesn't affect any functionalities. I will keep in mind for future reference

Reply:
You're correct that it's adequate for our needs. I was reviewing additional uses of the ontology.json. No further action is required. I will look into the RDF parser that Stanford wrote for similar use.